File: AwsEncryptionSDK/runtimes/rust/examples/keyring/aws_kms_hierarchical/version_branch_key_id.rs
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use aws_esdk::aws_cryptography_keyStore::client as keystore_client;
use aws_esdk::aws_cryptography_keyStore::types::key_store_config::KeyStoreConfig;
use aws_esdk::aws_cryptography_keyStore::types::KmsConfiguration;

/*
 This example demonstrates configuring a KeyStore and then
 uses a helper method to version a branch key.
*/
pub async fn version_branch_key_id(
    key_store_table_name: &str,
    logical_key_store_name: &str,
    kms_key_arn: &str,
    branch_key_id: &str
) -> Result<(), crate::BoxError> {
    // Create a Key Store
    // The KMS Configuration you use in the KeyStore MUST have the right access to the resources in the KeyStore.
    let sdk_config = aws_config::load_defaults(aws_config::BehaviorVersion::latest()).await;
    let key_store_config = KeyStoreConfig::builder()
        .kms_client(aws_sdk_kms::Client::new(&sdk_config))
        .ddb_client(aws_sdk_dynamodb::Client::new(&sdk_config))
        .ddb_table_name(key_store_table_name)
        .logical_key_store_name(logical_key_store_name)
        .kms_configuration(KmsConfiguration::KmsKeyArn(kms_key_arn.to_string()))
        .build()?;

    let keystore = keystore_client::Client::from_conf(key_store_config)?;

    // To version a branch key you MUST have access to kms:ReEncrypt* and kms:GenerateDataKeyWithoutPlaintext
    keystore.version_key()
        .branch_key_identifier(branch_key_id)
        .send()
        .await?;
    Ok(())
}

File: AwsEncryptionSDK/runtimes/go/examples/keyring/awskmshierarchicalkeyring/versionbranchkeyid.go
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package awskmshierarchicalkeyring

import (
	"context"
	"fmt"

	keystore "github.com/aws/aws-cryptographic-material-providers-library/releases/go/mpl/awscryptographykeystoresmithygenerated"
	keystoretypes "github.com/aws/aws-cryptographic-material-providers-library/releases/go/mpl/awscryptographykeystoresmithygeneratedtypes"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
	"github.com/aws/aws-sdk-go-v2/service/kms"
)

/*
This example demonstrates configuring a KeyStore and then
uses a helper method to version a branch key.
*/
func versionBranchKeyId(keyStoreTableName, logicalKeyStoreName, kmsKeyArn, branchKeyId string) error {
	// Load the AWS SDK configuration
	cfg, err := config.LoadDefaultConfig(context.Background())
	if err != nil {
		return err
	}
	// Create DDB and KMS clients
	ddbClient := dynamodb.NewFromConfig(cfg)
	kmsClient := kms.NewFromConfig(cfg)
	// Create the keystore
	// The KMS Configuration you use in the KeyStore MUST have the right access to the resources in the KeyStore.
	kmsConfig := keystoretypes.KMSConfigurationMemberkmsKeyArn{
		Value: kmsKeyArn,
	}
	keyStore, err := keystore.NewClient(keystoretypes.KeyStoreConfig{
		DdbTableName:        keyStoreTableName,
		KmsConfiguration:    &kmsConfig,
		LogicalKeyStoreName: logicalKeyStoreName,
		DdbClient:           ddbClient,
		KmsClient:           kmsClient,
	})
	if err != nil {
		return err
	}
	// To version a branch key you MUST have access to kms:ReEncrypt* and kms:GenerateDataKeyWithoutPlaintext
	_, err = keyStore.VersionKey(context.Background(), keystoretypes.VersionKeyInput{
		BranchKeyIdentifier: branchKeyId,
	})
	if err != nil {
		return err
	}
	return nil
}

// Function to test versionBranchKeyId in main.go in examples directory
func CreateAndVersionBranchKeyId(keyStoreKMSKeyRegion, keyStoreRegion, keyStoreKMSKeyID, keyStoreName, logicalKeyStoreName string) error {
	// Create the aws sdk clients
	cfg, err := config.LoadDefaultConfig(context.TODO())
	if err != nil {
		panic(err)
	}
	// Create the aws kms client
	kmsClient := kms.NewFromConfig(cfg, func(o *kms.Options) {
		o.Region = keyStoreKMSKeyRegion
	})
	// Create the ddb client
	ddbClient := dynamodb.NewFromConfig(cfg, func(options *dynamodb.Options) {
		options.Region = keyStoreRegion
	})
	// create branch key ID
	branchKeyId, err := createbranchkeyid(
		keyStoreName,
		logicalKeyStoreName,
		keyStoreKMSKeyID,
		ddbClient,
		kmsClient,
	)
	if err != nil {
		panic(err)
	}
	// Version Branch Key
	err = versionBranchKeyId(
		keyStoreName,
		logicalKeyStoreName,
		keyStoreKMSKeyID,
		branchKeyId,
	)
	if err != nil {
		panic(err)
	}
	fmt.Println("Create and version branchKey Id Example Completed Successfully")
	return nil
}
